### PR TITLE
Respect the Collection interface

### LINKED ISSUE #6: Chunk method does not work
I am experiencing a bug where the `chunk()` method for collections results in the error, `Missing argument 2 for Elasticquent\ElasticquentResultCollection::__construct(), called in /home/vagrant/code/aer-web/vendor/laravel/framework/src/Illuminate/Support/Collection.php on line 658 and defined`. 

I am using Laravel 5.1 and my code is as follows:
```php
$query = [
     'bool' => [
         'must' => [
             'multi_match' => [
                 'query' => $term,
                 'fields' => [
                     'title^5',
                     'blurb^3',
                     'description',
                     'location',
                     'zipcode',
                 ],
            ],
        ],
    ],
]; 
$campaigns = Campaign::searchByQuery($query);
dd($campaigns->chunk(2));
```

The full error is as follows:

```
ErrorException Missing argument 2 for Elasticquent\ElasticquentResultCollection::__construct(), called in /home/vagrant/code/aer-web/vendor/laravel/framework/src/Illuminate/Support/Collection.php on line 658 and defined
    vendor/elasticquent/elasticquent/src/ElasticquentResultCollection.php:21 Illuminate\Foundation\Bootstrap\HandleExceptions::handleError
    vendor/elasticquent/elasticquent/src/ElasticquentResultCollection.php:21 Elasticquent\ElasticquentResultCollection::__construct
    vendor/laravel/framework/src/Illuminate/Support/Collection.php:658 Illuminate\Support\Collection::chunk
    app/Http/Controllers/SearchController.php:48 App\Http\Controllers\SearchController::show
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Routing/Controller.php:256 Illuminate\Routing\Controller::callAction
    vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:164 Illuminate\Routing\ControllerDispatcher::call
    vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:112 Illuminate\Routing\ControllerDispatcher::Illuminate\Routing\{closure}
    [internal]:0 call_user_func
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:139 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    [internal]:0 call_user_func
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103 Illuminate\Pipeline\Pipeline::then
    vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:114 Illuminate\Routing\ControllerDispatcher::callWithinStack
    vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:69 Illuminate\Routing\ControllerDispatcher::dispatch
    vendor/laravel/framework/src/Illuminate/Routing/Route.php:201 Illuminate\Routing\Route::runWithCustomDispatcher
    vendor/laravel/framework/src/Illuminate/Routing/Route.php:134 Illuminate\Routing\Route::run
    vendor/laravel/framework/src/Illuminate/Routing/Router.php:704 Illuminate\Routing\Router::Illuminate\Routing\{closure}
    [internal]:0 call_user_func
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:139 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    [internal]:0 call_user_func
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103 Illuminate\Pipeline\Pipeline::then
    vendor/laravel/framework/src/Illuminate/Routing/Router.php:706 Illuminate\Routing\Router::runRouteWithinStack
    vendor/laravel/framework/src/Illuminate/Routing/Router.php:671 Illuminate\Routing\Router::dispatchToRoute
    vendor/laravel/framework/src/Illuminate/Routing/Router.php:631 Illuminate\Routing\Router::dispatch
    vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:229 Illuminate\Foundation\Http\Kernel::Illuminate\Foundation\Http\{closure}
    [internal]:0 call_user_func
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:139 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php:50 Illuminate\Foundation\Http\Middleware\VerifyCsrfToken::handle
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:124 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php:54 Illuminate\View\Middleware\ShareErrorsFromSession::handle
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:124 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:62 Illuminate\Session\Middleware\StartSession::handle
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:124 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php:37 Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::handle
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:124 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php:59 Illuminate\Cookie\Middleware\EncryptCookies::handle
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:124 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    app/Http/Middleware/CheckForMaintenanceMode.php:43 App\Http\Middleware\CheckForMaintenanceMode::handle
    [internal]:0 call_user_func_array
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:124 Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
    [internal]:0 call_user_func
    vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103 Illuminate\Pipeline\Pipeline::then
    vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:118 Illuminate\Foundation\Http\Kernel::sendRequestThroughRouter
    vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:86 Illuminate\Foundation\Http\Kernel::handle
    public/index.php:54 [main]
```

### PR DESCRIPTION
**Pull request by chuangbo**
Fix collection methods e.g. map, filter, each, pluck, etc.

Inspired from [this fork](https://github.com/clowdy/Elasticquent/blob/improvements/src/ElasticquentResultCollection.php#L20)
Implement #55 
Fixed #6 #12 #47 #49

### CODE REVIEW

**Comment by onbjerg** [comment]
I see what the issue with my implementation was now. This one is OK, but it introduces major breaking changes.

**Comment by chuangbo** [comment]
@onbjerg What's the major breaking changes? I should make it backwards compatible for this fix.

Also do we have any version cutting plans? @timgws

**Comment by onbjerg** [comment]
Well, so, the API changed. If you're battling around with e.g. ``ElasticquentResultCollection``, then the parameters are no longer the same. They've both changed.

Other than that, it's good. It should still appear as a major version bump in my opinion, though.

**Comment by chuangbo** [comment]
Oh you are right. `ElasticquentResultCollection` has to be changed. Shouldn't we encourage people to use `dev-master` in production?

**Comment by timgws** [comment]
Yeah, this is a breaking change.

> Shouldn't we encourage people to use `dev-master` in production?

Not really :) that being said, there are released versions here, and if you _are_ running dev-master in production, then you are basically signing away your rights to a stable API.

That being said, I don't feel like we should do a major version bump just for this item.

Moving forward, I think you should try and make this backwards compatible for the moment, mark it with a `@todo`, and we will do the breaking change a little further down the line when we have more reasons to break the API.

**Comment by chuangbo** [comment]
@timgws OK I'll try to make `__contructor` compatible. Anything else?

**Comment by chuangbo** [comment]
Backwards compatible is done.

Now we allow to generate ResultCollection by either:

New way, which I've changed in readme:
```php
$collection = Book::hydrateElasticsearchResult($client->search($params));
```

Old way, with a warning log:
```php
$collection = new \Elasticquent\ElasticquentResultCollection($client->search($params), new Book);

// Warning in /storage/logs/laravel.log
// WARNING: Deprecated: please use Book::hydrateElasticsearchResult($result).
```

**Comment by timgws** [comment]
Please don't use facade's inside libraries like this. The dependancy should be injected.

Also, putting an item like this in a log file is a really bad idea, and should be avoided.

**Comment by chuangbo** [comment]
Fixed and thank you for letting me know :smile:

File: src/ElasticquentResultCollection.php [hunk 1]
```diff
@@ -47,11 +47,11 @@ public function __construct($items, $meta = null)
      */
     public function setMeta(array $meta)
     {
-        $this->took = data_get($meta, 'took');
-        $this->timed_out = data_get($meta, 'timed_out');
-        $this->shards = data_get($meta, '_shards');
-        $this->hits = data_get($meta, 'hits');
-        $this->aggregations = data_get($meta, 'aggregations', []);
+        $this->took = $meta['took'];
```
Comment (timgws) [inline src/ElasticquentResultCollection.php]: Can we change this to something like

```
$this->took = isset($meta['took']) ? $meta['took'] : null;
```

That way, at least if the given metadata items are not correctly set, then this code will still continue to work.

Comment (chuangbo) [inline src/ElasticquentResultCollection.php]: Done

**Comment by chuangbo** [comment]
Thank you!

**Pull request merged**